apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/FlushableConsumerBase.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright 2017 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Diagnostics.Common
+{
+    /// <summary>
+    /// Base class for the <see cref="IFlushableConsumer{T}"/> that exposes a <see cref="SemaphoreSlim"/> to
+    /// protect the buffer.
+    /// </summary>
+    internal abstract class FlushableConsumerBase<T> : IFlushableConsumer<T>
+    {
+        /// <summary>A semaphore to protect the buffer.</summary>
+        protected readonly SemaphoreSlim Semaphore = new SemaphoreSlim(1);
```

Comment:
I think it can be simpler than that, actually - I think the semaphore can be entirely private, with just abstract methods of

- ReceiveWithSemaphoreHeld
- FlushWithSemaphoreHeld

(and async versions). The base class can do all the semaphore handling, which should really simplify the concrete subclasses.